
# sciL3Pipe

A Snakemake pipeline for processing and aligning sci-L3-seq data.


Required packages:

* Minimum required `snakemake 9.11.6`
* `snakemake-executor-plugin-cluster-generic`
* For hoffman2 specific executor:
[snakemake-executor-plugin-cluster-hoffman2](https://github.com/peterch405/snakemake-executor-plugin-cluster-hoffman2)




## Running pipeline from raw fastq files:

1. (Optional) Create a folder with symbolic links to fastq files
    ```mkdir fastqs```
    ```ln -s <path to raw fastq> <path to fastqs folder>```

2. Create sample.json file that informs the pipeline which fastq files correspond to which sample:
    ```python fastq2json.py --fastq_dir tests/fastqs/```

    Will create a file with the following structure:
    ```
    {
    "yi293": {
        "R1": [
            "/u/project/yeastyin/chovanec/lib293/fastqs/yi293_S7_R1_001.fastq.gz"
        ],
        "R2": [
            "/u/project/yeastyin/chovanec/lib293/fastqs/yi293_S7_R2_001.fastq.gz"
        ]
        }
    }
    ```

3. Create a config.yaml file specific to your run. The config file contains pipeline input parameters. Example config that can be run to test files in included `configs/config.yaml` in this repo:

```
################################################################################
# Input/Output
################################################################################
# output directory (include trailing /)
output_dir: "tests/"
# temporary directory for intermediate files
temp_dir: "tests/temp/"
# fastq file location produced by fastq2json.py
fastqs: "tests/samples.json"
# skip splitting input fastq files into smaller chunks to speed up processing
split_lines: 1000000
################################################################################
# SSS extract and split
################################################################################
# samples with corresponding SSS barcode file, must match fastqs
samples:
    yi401: "workflow/resources/barcodes/barcode_sss_401_nova.txt"
sss_mismatches: 0
rt_mismatches: 3
rt_primer: GGGATGCAGCTCGCTCCTG
################################################################################
# Aligner settings
################################################################################
aligner: "bwa"
assembly: "hg38"
# assembly: "hg38_test"
# start from trimmed fastq, not requiring prior files, file generated by trimmed2json.py
from_trimmed: "tests/trimmed.json"
################################################################################
# Tn5 and ligation barcode
################################################################################
# the round1 barcode list (on tn5) for splitting single cells
tn5_barcodes: "workflow/resources/barcodes/barcode_tn5.txt"
# the round2 barcode list (by ligation) for splitting single cells
ligation_barcodes: "workflow/resources/barcodes/barcode_ligation_104.txt"
tn5_mismatches: 1
ligation_mismatches: 1
# Reads per single cell to keep (default: 10000)
cutoff: 2000
################################################################################   
# Indexes
################################################################################
hybrid_reference: False
bwa_index:
    hg38: "/genomes/GRCh38/GRCh38"
bowtie2_index:
    hg38: "/genomes/bowtie2_indexes/GRCh38"
################################################################################
# BAM split into single cells
################################################################################
# perform splitting of SSS BAMs into single cell BAMs
split_bam: True
# SSS bam locations to split, produced by sss_bam2json.py
# "tests/test_BAM/sss_samples.json"
sss_bams: "tests/sss_samples.json"
################################################################################
# Summary metrics
################################################################################
# generate all files for summary metrics/QC, requires split_bam to have been run
generate_summary: True
# generate summary for sequences with prefix "wprefix" (chr prefix: chr1, chr2) or human without prefix "woprefix" (no prefix: 1,2,...X,Y)
summary_for: "woprefix"
run_collisions: False
################################################################################
# Hybrid BAM split
################################################################################
# Will only be run if BAM filter also run.
run_split: True
# chr assembly prefix
prefix_1: "mouse"
# non-chr assembly prefix
prefix_2: "human"
hybrid_filter: "NC_007605,hs37d5"
################################################################################
# BAM filter
################################################################################
run_filter: True
natsort_chromosomes: True
# If hybrid assembly, specify which to process, else can use genome assembly as a postfix
process_only: "human"
# If not running hybrid split, can specify filter here instead of hybrid_filter
additional_filter: "hs37d5"
################################################################################
# Sample assignment
################################################################################
sample_assignment:
    - H9_primed:
        lib_SSS:
            - yi401_CGCTTG
            - yi401_GTCCTA
        Tn5:
            - "TGATATTG"
            - "GATCCCGT"
            - "CTCGATTA"
            - "CATCAAGG"
            - "TCCTTGTG"
            - "GGTCATAT"
            - "ATCGCGTT"
    - H9_primed:
        lib_SSS:
            - yi401_ACGCGA
        Tn5:
            - "TGATATTG"
            - "GATCCCGT"
            - "CTCGATTA"
            - "CATCAAGG"
            - "TCCTTGTG"
            - "GGTCATAT"
            - "ATCGCGTT"
    - WIBR3:
        lib_SSS:
            - yi401_ACGCGA
            - yi401_CGCTTG
            - yi401_GTCCTA
        Tn5:
            - "CATGCCCC"
            - "GTTACGCG"
            - "CCGCGCTT"
            - "TCTTAGTG"
            - "TCGGCCTA"
            - "CTTTCTCT"
            - "TCGCGTTT"
    - LNS:
        lib_SSS:
            - yi401_ACGCGA
            - yi401_CGCTTG
            - yi401_GTCCTA
        Tn5:
            - "GTCAGTAG"
            - "CCATGGAA"
            - "ATGCTGCG"
            - "GAGTCTTT"
            - "TACGATAT"
            - "ACCATTTA"
            - "ATCGGGAC"
            - "GACGTCGG"
    - BJ5ta:
        lib_SSS:
            - yi401_ACGCGA
            - yi401_CGCTTG
            - yi401_GTCCTA
        Tn5:
            - "CATTGTGT"
            - "TTTGACTC"
```

1. To run the pipeline:

```
snakemake --configfile configs/config.yaml
```

Conveniance shell script is included `run_pipeline_local.sh` and `run_pipeline_sge.sh`


The pipeline uses docker container which is automatically pulled from dockerhub. Note, apptainer needs to be installed.


## Running pipeline from trimmed fastq or SSS BAM files:

The pipeline can alternatively be started from trimmed fastq or SSS BAM files. 

To start from trimmed fastq files:
1. Create trimmed.json file that informs the pipeline which fastq files correspond to which sample:

    ```python trimmed2json.py --fastq_dir tests/trimmed_fastq/```

    Similarly for starting from SSS BAMs:

    ```python sss_bam2json.py --bam_dir tests/bwa_alignment/```



2. Add the path to the json file in the `config.yaml`:

```
# start from trimmed fastq, not requiring prior files, file generated by trimmed2json.py
from_trimmed: "tests/trimmed.json"
```

```
# SSS bam locations to split, produced by sss_bam2json.py
sss_bams: "tests/sss_samples.json"
```

## QC output

Summary of reads for each single cells:
`genome_coverage_bwa/single_cells/single_cell_summary_all.txt`

| Cell ID | Mapped reads (counts R1 and R2, double counting) | Mapped reads MAPQ>0 | Unique reads by R1 start and end | Mapped reads MAPQ>0 over R1 start and end (IVT over Tn5 ratio) |
| :---                        | :--: | :--: | :--: |  ---:  |
|yy409_TTTCGC.TGATATTGGGGTACA |40376 |17783 |15879 |1.11991 |
|yy409_TTTCGC.TGATATTGGCTTGAT |26650 |11328 |10137 |1.11749 |
|yy409_TTTCGC.TGATATTGAGAGACT |10055 |4714  |4169  |1.13073 |
|yy409_TTTCGC.TGATATTGCACTGTT |76566 |33888 |30170 |1.12324 |
|yy409_TTTCGC.TGATATTGATAAAAG |19368 |8284  |7374  |1.12341 |


Barcodes in BAM read names:
`GAGTCTTT,TAGTCTA,GCTTG,TGTA,H00910:58:AACYNFGM5:1:2204:15656:40340`

| Tn5 | Ligation | SSS | UMI |
| :--- | :--: | :--: |  ---:  |
| GAGTCTTT| GAGGCGA | CGCTTG | TTCT |


# Report

A snakemake html report with QC and logs an be obtained by running:
```
snakemake --configfile configs/config.yaml --report report.html
```
